### PR TITLE
[CLEANUP] Prune terraform.NewResourceConfig

### LINKED ISSUE #22276: Habitat Provisioner Crash
### Terraform Version
```
❯ terraform -v
Terraform v0.12.7-dev
+ provider.aws v2.21.1
```

### Terraform Configuration Files
```hcl
provider "aws" {
  region = "us-east-1"
}

resource "aws_instance" "redis" {
  count = 3
  instance_type = "t2.micro"
  ami = "xxx"

  provisioner "habitat" {
    peer = "${aws_instance.redis.0.private_ip}"
    use_sudo = true
    service_type = "systemd"

    service {
      name = "core/redis"
      topology = "leader"
      user_toml = "${file("conf/redis.toml")}"
    }
  }
}
```

### Crash Output
https://gist.github.com/appilon/10c62ec1d801d43343782568faa31a4c

### Steps to Reproduce
1. `terraform init`
2. `terraform validate`

### Additional Context
This was discovered during a larger PR pruning some old code from core. A fix will come along in that PR. The reason for it is the shim layers supplying `[]interface{}{ map[string]interface{} }` when the provisioner tries to cast directly to `[]map[string]interface{}`

### References
#22270 
HCL1 apparently could create a `[]map[string]interface`: https://github.com/hashicorp/hcl/blob/cf7d376da96d9cecec7c7483cec2735efe54a410/decoder.go#L245

### PR DESCRIPTION
**Pull request by appilon**
Resurrected from https://github.com/hashicorp/terraform/pull/21601. The switch to the shims required a few changes all in the same vein, do not cast directly to `[]map[string]interface` but rather first `[]interface` then `map[string]interface{}`. Apologies for anything deleted that should not be, happy to restore anything.

Closes #22276

### CODE REVIEW

File: helper/schema/schema_test.go [hunk 1]
```diff
@@ -5329,6 +5329,7 @@ func TestSchemaMap_Validate(t *testing.T) {
 				},
 			},
 		},
+		*/
```
Comment (teamterraform) [inline helper/schema/schema_test.go]: Out of curiosity, does this test pass if you add `Err: True`? I'm wondering if this is one of those cases where HCL2 (or some other update) is helpfully catching the error sooner than before (but not curious enough to test locally ;) )

File: terraform/resource.go [hunk 2]
```diff
@@ -199,15 +184,32 @@ type ResourceConfig struct {
...
+// something is computed, but it must not contain unprocessed interpolation
+// sequences as we might've seen in Terraform v0.11 and prior.
+func NewResourceConfigRaw(raw map[string]interface{}) *ResourceConfig {
+	v := hcl2shim.HCL2ValueFromConfigValue(raw)
+
+	// This is a little weird but we round-trip the value through the hcl2shim
+	// package here for two reasons: firstly, because that reduces the risk
+	// of it including something unlike what NewResourceConfigShimmed would
+	// produce, and secondly because it creates a copy of "raw" just in case
+	// something is relying on the fact that in the old world thw raw and
```
Comment (kmoe) [inline terraform/resource.go]:
```suggestion
	// something is relying on the fact that in the old world the raw and
```

**Comment by jbardin** [comment]
The new switch case looks sufficient to handle the legacy configs that might come in. 👍

**Comment by appilon** [comment]
the failures were related to some github errors while fetching the modules, tests are green

**Pull request merged**